File: scripts/decompile.py
# Jython script to decompile all functions in an imported binary via Headless Analyzer
# @category Decompiler

from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor
import os

def run():
    program = currentProgram
    if not program:
        print("[-] Error: No program loaded.")
        return

    print("[+] Loaded target program: %s" % program.getName())
    
    decomp_interface = DecompInterface()
    decomp_interface.openProgram(program)
    
    fm = program.getFunctionManager()
    funcs = fm.getFunctions(True) # True iterates functions in memory address order
    
    output_file_path = "decompiled_output.c"
    print("[+] Commencing decompilation. Output file: %s" % output_file_path)
    
    count = 0
    with open(output_file_path, "w") as f:
        f.write("/**\n")
        f.write(" * Automatically decompiled using Ghidra Headless Analyzer\n")
        f.write(" * Target Binary: %s\n" % program.getName())
        f.write(" */\n\n")
        
        for func in funcs:
            count += 1
            f.write("// =========================================\n")
            f.write("// Function: %s at 0x%s\n" % (func.getName(), func.getEntryPoint()))
            f.write("// =========================================\n")
            
            # Decompile each function with a 30-second execution safety limit
            results = decomp_interface.decompileFunction(func, 30, ConsoleTaskMonitor())
            if results.decompileCompleted():
                decompiled_code = results.getDecompiledFunction().getC()
                if decompiled_code:
                    f.write(decompiled_code)
                else:
                    f.write("// [Warning] Decompilation succeeded but C output is empty.\n")
            else:
                f.write("// [Error] Failed to decompile %s: %s\n" % (func.getName(), results.getErrorMessage()))
            f.write("\n\n")
            
    print("[+] Decompilation process complete. Parsed %d functions." % count)

if __name__ == "__main__":
    run()

File: scripts/decompile.java
// Native Java script to decompile functions in chunks of 500 and output to .txt files
// @category Decompiler

import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.util.task.ConsoleTaskMonitor;
import java.io.FileWriter;
import java.io.PrintWriter;

public class decompile extends GhidraScript {
    @Override
    public void run() throws Exception {
        if (currentProgram == null) {
            println("[-] Error: No program loaded.");
            return;
        }

        println("[+] Loaded target program: " + currentProgram.getName());

        DecompInterface decompInterface = new DecompInterface();
        decompInterface.openProgram(currentProgram);

        FunctionIterator funcs = currentProgram.getFunctionManager().getFunctions(true);

        int chunkIndex = 1;
        int count = 0;
        PrintWriter writer = null;

        println("[+] Commencing chunked decompilation (500 functions per file)... ");

        try {
            while (funcs.hasNext() && !monitor.isCancelled()) {
                Function func = funcs.next();
                
                // Trigger file rotation at the start and after every 500 functions
                if (count % 500 == 0) {
                    if (writer != null) {
                        writer.close();
                    }
                    String fileName = "decompiled_chunk_" + chunkIndex + ".txt";
                    println("[+] Opening chunk file: " + fileName);
                    writer = new PrintWriter(new FileWriter(fileName));
                    writer.println("/**");
                    writer.println(" * Automatically decompiled using Ghidra Headless Analyzer (Java Engine)");
                    writer.println(" * Target Binary: " + currentProgram.getName());
                    writer.println(" * Chunk Index: " + chunkIndex);
                    writer.println(" */\n");
                    chunkIndex++;
                }

                count++;
                writer.println("// =========================================");
                writer.println("// Function: " + func.getName() + " at 0x" + func.getEntryPoint());
                writer.println("// =========================================");

                DecompileResults results = decompInterface.decompileFunction(func, 30, new ConsoleTaskMonitor());
                if (results.decompileCompleted()) {
                    String decompiledCode = results.getDecompiledFunction().getC();
                    if (decompiledCode != null && !decompiledCode.trim().isEmpty()) {
                        writer.println(decompiledCode);
                    } else {
                        writer.println("// [Warning] Decompilation succeeded but C output is empty.");
                    }
                } else {
                    writer.println("// [Error] Failed to decompile " + func.getName() + ": " + results.getErrorMessage());
                }
                writer.println("\n");
            }
        } finally {
            if (writer != null) {
                writer.close();
            }
        }

        println("[+] Decompilation process complete. Generated " + (chunkIndex - 1) + " .txt chunk files (Total: " + count + " functions).");
    }
}
